README.md
# Jetscii

A tiny library to efficiently search strings for sets of ASCII
characters or byte slices for sets of bytes.

[![Current Version](https://img.shields.io/crates/v/jetscii.svg)](https://crates.io/crates/jetscii) [![Documentation](https://docs.rs/jetscii/badge.svg)][docs]

## Examples

### Searching for a set of ASCII characters

```rust
use jetscii::ascii_chars;

fn main() {
    let part_number = "86-J52:rev1";
    let first = ascii_chars!('-', ':').find(part_number);
    assert_eq!(first, Some(2));
}
```

### Searching for a set of bytes

```rust
use jetscii::bytes;

fn main() {
    let raw_data = [0x00, 0x01, 0x10, 0xFF, 0x42];
    let first = bytes!(0x01, 0x10).find(&raw_data);
    assert_eq!(first, Some(1));
}
```

Check out [the documentation][docs] for information about feature flags and benchmarks.

[docs]: https://docs.rs/jetscii/

## Contributing

1. Fork it (https://github.com/shepmaster/jetscii/fork)
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Add a failing test.
4. Add code to pass the test.
5. Commit your changes (`git commit -am 'Add some feature'`)
6. Ensure tests pass.
7. Push to the branch (`git push origin my-new-feature`)
8. Create a new Pull Request


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][] and this project adheres
to [Semantic Versioning][].

[Keep a Changelog]: https://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: https://semver.org/spec/v2.0.0.html

## [Unreleased]

- Trait bounds removed from generic types, bounds are only required for impls

## [0.5.3] - 2022-07-06

- Fix buffer overflows in find. (#55)

## [0.5.2] - 2022-05-08

- Support 32-bit code with SSE4.2. (#52)

## [0.5.1] - 2021-09-18

- Avoid accessing invalid memory when the search string isn't found
  and the haystack ends on a OS memory page. (#45)

## [0.5.0] - 2021-05-05

- Update minimum required Rust version to Rust 1.51. (#39)

## [0.4.3] - 2018-08-26

- Coalesce back-to-back instructions for small performance
  improvements. (#35)

## [0.4.3] - 2018-08-26

### Fixed

- Coalesce back-to-back instructions for small performance
  improvements. (#24)

### Changes

 - Library is now dual-licensed as MIT or APACHE-2. (#12)

## [0.4.2] - 2018-07-11

### Fixed

- No longer reports false positives in certain cases when input data
  is not aligned on a 16-byte boundary. (#19)

## [0.4.1] - 2018-07-07

### Added

- `Substring` support restored. (#20)

## [0.4.0] - 2018-05-13

### Changed

- Complete rewrite from inline assembly to stable compiler intrinsics.

### Added

- Directly searching through a slice of bytes is now supported.

### Removed

- `Substring` was removed during the rewrite.


src/lib.rs
#![cfg_attr(feature = "pattern", feature(pattern))]

//! A tiny library to efficiently search strings for sets of ASCII
//! characters or byte slices for sets of bytes.
//!
//! ## Examples
//!
//! ### Searching for a set of ASCII characters
//!
//! ```rust
//! use jetscii::ascii_chars;
//!
//! fn main() {
//!     let part_number = "86-J52:rev1";
//!     let first = ascii_chars!('-', ':').find(part_number);
//!     assert_eq!(first, Some(2));
//! }
//! ```
//!
//! ### Searching for a set of bytes
//!
//! ```rust
//! use jetscii::bytes;
//!
//! fn main() {
//!     let raw_data = [0x00, 0x01, 0x10, 0xFF, 0x42];
//!     let first = bytes!(0x01, 0x10).find(&raw_data);
//!     assert_eq!(first, Some(1));
//! }
//! ```
//!
//! ### Searching for a substring
//!
//! ```
//! use jetscii::Substring;
//!
//! let colors = "red, blue, green";
//! let first = Substring::new(", ").find(colors);
//! assert_eq!(first, Some(3));
//! ```
//!
//! ### Searching for a subslice
//!
//! ```
//! use jetscii::ByteSubstring;
//!
//! let raw_data = [0x00, 0x01, 0x10, 0xFF, 0x42];
//! let first = ByteSubstring::new(&[0x10, 0xFF]).find(&raw_data);
//! assert_eq!(first, Some(2));
//! ```
//!
//! ## Using the pattern API
//!
//! If this crate is compiled with the unstable `pattern` feature
//! flag, [`AsciiChars`] will implement the
//! [`Pattern`][std::str::pattern::Pattern] trait, allowing it to be
//! used with many traditional methods.
//!
//! ```
//! # #[cfg(feature = "pattern")]
//! use jetscii::ascii_chars;
//!
//! fn main() {
//! # #[cfg(feature = "pattern")] {
//!     let part_number = "86-J52:rev1";
//!     let parts: Vec<_> = part_number.split(ascii_chars!('-', ':')).collect();
//!     assert_eq!(&parts, &["86", "J52", "rev1"]);
//! # }
//! }
//! ```
//!
//! ```
//! # #[cfg(feature = "pattern")] {
//! use jetscii::Substring;
//!
//! let colors = "red, blue, green";
//! let colors: Vec<_> = colors.split(Substring::new(", ")).collect();
//! assert_eq!(&colors, &["red", "blue", "green"]);
//! # }
//! ```
//!
//! ## What's so special about this library?
//!
//! We use a particular set of SSE 4.2 instructions (`PCMPESTRI`
//! and `PCMPESTRM`) to gain great speedups. This method stays fast even
//! when searching for a byte in a set of up to 16 choices.
//!
//! When the `PCMPxSTRx` instructions are not available, we fall back to
//! reasonably fast but universally-supported methods.
//!
//! ## Benchmarks
//!
//! These numbers come from running on my personal laptop; always
//! benchmark with data and machines similar to your own.
//!
//! ### Single character
//!
//! Searching a 5MiB string of `a`s with a single space at the end for a space:
//!
//! | Method                                                      | Speed          |
//! |-------------------------------------------------------------|----------------|
//! | <code>ascii_chars!(' ').find(s)</code>                      | 11504 MB/s     |
//! | <code>s.as_bytes().iter().position(\|&c\| c == b' ')</code> | 2377 MB/s      |
//! | <code>s.find(" ")</code>                                    | 2149 MB/s      |
//! | <code>s.find(&[' '][..])</code>                             | 1151 MB/s      |
//! | **<code>s.find(' ')</code>**                                | **14600 MB/s** |
//! | <code>s.find(\|c\| c == ' ')</code>                         | 1080 MB/s      |
//!
//! ### Set of 3 characters
//!
//! Searching a 5MiB string of `a`s with a single ampersand at the end for `<`, `>`, and `&`:
//!
//! | Method                                                      | Speed          |
//! |-------------------------------------------------------------|----------------|
//! | **<code>ascii_chars!(/\* ... \*/).find(s)</code>**          | **11513 MB/s** |
//! | <code>s.as_bytes().iter().position(\|&c\| /* ... */)</code> | 1644 MB/s      |
//! | <code>s.find(&[/* ... */][..])</code>                       | 1079 MB/s      |
//! | <code>s.find(\|c\| /* ... */))</code>                       | 1084 MB/s      |
//!
//! ### Set of 5 characters
//!
//! Searching a 5MiB string of `a`s with a single ampersand at the end for `<`, `>`, `&`, `'`, and `"`:
//!
//! | Method                                                      | Speed          |
//! |-------------------------------------------------------------|----------------|
//! | **<code>ascii_chars!(/\* ... \*/).find(s)</code>**          | **11504 MB/s** |
//! | <code>s.as_bytes().iter().position(\|&c\| /* ... */)</code> | 812 MB/s       |
//! | <code>s.find(&[/* ... */][..]))</code>                      | 538 MB/s       |
//! | <code>s.find(\|c\| /* ... */)</code>                        | 1082 MB/s      |
//!
//! ### Substring
//!
//! Searching a 5MiB string of `a`s with the string "xyzzy" at the end for "xyzzy":
//!
//! | Method                                           | Speed          |
//! |--------------------------------------------------|----------------|
//! | **<code>Substring::new("xyzzy").find(s)</code>** | **11475 MB/s** |
//! | <code>s.find("xyzzy")</code>                     | 5391 MB/s      |

use std::marker::PhantomData;

include!(concat!(env!("OUT_DIR"), "/src/macros.rs"));

#[cfg(any(jetscii_sse4_2 = "yes", jetscii_sse4_2 = "maybe"))]
mod simd;

#[cfg(any(jetscii_sse4_2 = "maybe", jetscii_sse4_2 = "no"))]
mod fallback;

#[cfg(feature = "pattern")]
mod pattern;

macro_rules! dispatch {
    (simd: $simd:expr,fallback: $fallback:expr,) => {
        // If we can tell at compile time that we have support,
        // call the optimized code directly.
        #[cfg(jetscii_sse4_2 = "yes")]
        {
            $simd
        }

        // If we can tell at compile time that we will *never* have
        // support, call the fallback directly.
        #[cfg(jetscii_sse4_2 = "no")]
        {
            $fallback
        }

        // Otherwise, we will be run on a machine with or without
        // support, so we perform runtime detection.
        #[cfg(jetscii_sse4_2 = "maybe")]
        {
            if is_x86_feature_detected!("sse4.2") {
                $simd
            } else {
                $fallback
            }
        }
    };
}

/// Searches a slice for a set of bytes. Up to 16 bytes may be used.
pub struct Bytes<F> {
    #[cfg(any(jetscii_sse4_2 = "yes", jetscii_sse4_2 = "maybe"))]
    simd: simd::Bytes,

    #[cfg(any(jetscii_sse4_2 = "maybe", jetscii_sse4_2 = "no"))]
    fallback: fallback::Bytes<F>,

    // Since we might not use the fallback implementation, we add this
    // to avoid unused type parameters.
    _fallback: PhantomData<F>,
}

impl<F> Bytes<F>
where
    F: Fn(u8) -> bool,
{
    /// Manual constructor; prefer using [`bytes!`] instead.
    ///
    /// Provide an array of bytes to search for, the number of
    /// valid bytes provided, and a closure to use when the SIMD
    /// intrinsics are not available. The closure **must** search for
    /// the same bytes as in the array.
    #[allow(unused_variables)]
    #[must_use]
    pub fn new(bytes: [u8; 16], len: i32, fallback: F) -> Self {
        Bytes {
            #[cfg(any(jetscii_sse4_2 = "yes", jetscii_sse4_2 = "maybe"))]
            simd: simd::Bytes::new(bytes, len),

            #[cfg(any(jetscii_sse4_2 = "maybe", jetscii_sse4_2 = "no"))]
            fallback: fallback::Bytes::new(fallback),

            _fallback: PhantomData,
        }
    }

    /// Searches the slice for the first matching byte in the set.
    #[inline]
    #[must_use]
    pub fn find(&self, haystack: &[u8]) -> Option<usize> {
        dispatch! {
            simd: unsafe { self.simd.find(haystack) },
            fallback: self.fallback.find(haystack),
        }
    }
}

/// A convenience type that can be used in a constant or static.
pub type BytesConst = Bytes<fn(u8) -> bool>;

/// Searches a string for a set of ASCII characters. Up to 16
/// characters may be used.
pub struct AsciiChars<F>(Bytes<F>);

impl<F> AsciiChars<F>
where
    F: Fn(u8) -> bool,
{
    /// Manual constructor; prefer using [`ascii_chars!`] instead.
    ///
    /// Provide an array of ASCII bytes to search for, the number of
    /// valid bytes provided, and a closure to use when the SIMD
    /// intrinsics are not available. The closure **must** search for
    /// the same characters as in the array.
    ///
    /// ### Panics
    ///
    /// - If you provide a non-ASCII byte.
    #[must_use]
    pub fn new(chars: [u8; 16], len: i32, fallback: F) -> Self {
        for &b in &chars {
            assert!(b < 128, "Cannot have non-ASCII bytes");
        }
        Self(Bytes::new(chars, len, fallback))
    }

    /// Searches the string for the first matching ASCII byte in the set.
    #[inline]
    #[must_use]
    pub fn find(&self, haystack: &str) -> Option<usize> {
        self.0.find(haystack.as_bytes())
    }
}

/// A convenience type that can be used in a constant or static.
pub type AsciiCharsConst = AsciiChars<fn(u8) -> bool>;

/// Searches a slice for the first occurence of the subslice.
pub struct ByteSubstring<'a> {
    #[cfg(any(jetscii_sse4_2 = "yes", jetscii_sse4_2 = "maybe"))]
    simd: simd::ByteSubstring<'a>,

    #[cfg(any(jetscii_sse4_2 = "maybe", jetscii_sse4_2 = "no"))]
    fallback: fallback::ByteSubstring<'a>,
}

impl<'a> ByteSubstring<'a> {
    #[must_use]
    pub fn new(needle: &'a [u8]) -> Self {
        ByteSubstring {
            #[cfg(any(jetscii_sse4_2 = "yes", jetscii_sse4_2 = "maybe"))]
            simd: simd::ByteSubstring::new(needle),

            #[cfg(any(jetscii_sse4_2 = "maybe", jetscii_sse4_2 = "no"))]
            fallback: fallback::ByteSubstring::new(needle),
        }
    }

    #[cfg(feature = "pattern")]
    #[must_use]
    fn needle_len(&self) -> usize {
        dispatch! {
            simd: self.simd.needle_len(),
            fallback: self.fallback.needle_len(),
        }
    }

    /// Searches the slice for the first occurence of the subslice.
    #[inline]
    #[must_use]
    pub fn find(&self, haystack: &[u8]) -> Option<usize> {
        dispatch! {
            simd: unsafe { self.simd.find(haystack) },
            fallback: self.fallback.find(haystack),
        }
    }
}

/// A convenience type that can be used in a constant or static.
pub type ByteSubstringConst = ByteSubstring<'static>;

/// Searches a string for the first occurence of the substring.
pub struct Substring<'a>(ByteSubstring<'a>);

impl<'a> Substring<'a> {
    #[must_use]
    pub fn new(needle: &'a str) -> Self {
        Substring(ByteSubstring::new(needle.as_bytes()))
    }

    #[cfg(feature = "pattern")]
    #[must_use]
    fn needle_len(&self) -> usize {
        self.0.needle_len()
    }

    /// Searches the string for the first occurence of the substring.
    #[inline]
    #[must_use]
    pub fn find(&self, haystack: &str) -> Option<usize> {
        self.0.find(haystack.as_bytes())
    }
}

/// A convenience type that can be used in a constant or static.
pub type SubstringConst = Substring<'static>;
